From 2306733941463a1de2d12f960c7f89fb8623d3b7 Mon Sep 17 00:00:00 2001
From: Daiki Ueno <ueno@gnu.org>
Date: Thu, 26 Feb 2026 08:06:07 +0900
Subject: [PATCH 04/11] BACKPORT: accelerated: give up on defining nettle HMAC
 interface

Nettle 4 doesn't provide a way to define custom HMAC instances.

Signed-off-by: Daiki Ueno <ueno@gnu.org>
(cherry picked from commit 4e3921c36529110a94c2a63e0d6601c502901589)
Signed-off-by: Xi Ruoyao <xry111@xry111.site>
---
 lib/accelerated/aarch64/hmac-sha-aarch64.c |  4 ++--
 lib/accelerated/x86/hmac-padlock.c         |  4 ++--
 lib/accelerated/x86/hmac-x86-ssse3.c       |  4 ++--
 lib/accelerated/x86/x86-common.c           | 15 +++++++++++++++
 4 files changed, 21 insertions(+), 6 deletions(-)

diff --git a/lib/accelerated/aarch64/hmac-sha-aarch64.c b/lib/accelerated/aarch64/hmac-sha-aarch64.c
index 6261180f0..c36f4fe03 100644
--- a/lib/accelerated/aarch64/hmac-sha-aarch64.c
+++ b/lib/accelerated/aarch64/hmac-sha-aarch64.c
@@ -35,7 +35,7 @@
 #include "sha-aarch64.h"
 #include "algorithms.h"
 
-#ifdef HAVE_LIBNETTLE
+#if defined(HAVE_LIBNETTLE) && defined(HMAC_SET_KEY)
 
 typedef void (*update_func)(void *, size_t, const uint8_t *);
 typedef void (*digest_func)(void *, size_t, uint8_t *);
@@ -301,4 +301,4 @@ const gnutls_crypto_mac_st _gnutls_hmac_sha_aarch64 = {
 	.fast = wrap_aarch64_hmac_fast,
 };
 
-#endif /* HAVE_LIBNETTLE */
+#endif /* HAVE_LIBNETTLE && HMAC_SET_KEY */
diff --git a/lib/accelerated/x86/hmac-padlock.c b/lib/accelerated/x86/hmac-padlock.c
index ac91d20ec..b4e16e273 100644
--- a/lib/accelerated/x86/hmac-padlock.c
+++ b/lib/accelerated/x86/hmac-padlock.c
@@ -36,7 +36,7 @@
 #include "sha-padlock.h"
 #include "algorithms.h"
 
-#ifdef HAVE_LIBNETTLE
+#if defined(HAVE_LIBNETTLE) && defined(HMAC_SET_KEY)
 
 #define IPAD 0x36
 #define OPAD 0x5c
@@ -359,4 +359,4 @@ const gnutls_crypto_mac_st _gnutls_hmac_sha_padlock = {
 	.fast = wrap_padlock_hmac_fast,
 };
 
-#endif /* HAVE_LIBNETTLE */
+#endif /* HAVE_LIBNETTLE && HMAC_SET_KEY */
diff --git a/lib/accelerated/x86/hmac-x86-ssse3.c b/lib/accelerated/x86/hmac-x86-ssse3.c
index dec9c2688..61b1ea864 100644
--- a/lib/accelerated/x86/hmac-x86-ssse3.c
+++ b/lib/accelerated/x86/hmac-x86-ssse3.c
@@ -35,7 +35,7 @@
 #include "sha-x86.h"
 #include "algorithms.h"
 
-#ifdef HAVE_LIBNETTLE
+#if defined(HAVE_LIBNETTLE) && defined(HMAC_SET_KEY)
 
 typedef void (*update_func)(void *, size_t, const uint8_t *);
 typedef void (*digest_func)(void *, size_t, uint8_t *);
@@ -297,4 +297,4 @@ const gnutls_crypto_mac_st _gnutls_hmac_sha_x86_ssse3 = {
 	.fast = wrap_x86_hmac_fast,
 };
 
-#endif /* HAVE_LIBNETTLE */
+#endif /* HAVE_LIBNETTLE && HMAC_SET_KEY */
diff --git a/lib/accelerated/x86/x86-common.c b/lib/accelerated/x86/x86-common.c
index aaa74782c..db625ac71 100644
--- a/lib/accelerated/x86/x86-common.c
+++ b/lib/accelerated/x86/x86-common.c
@@ -35,6 +35,7 @@
 #include "x86-common.h"
 #ifdef HAVE_LIBNETTLE
 #include <nettle/aes.h> /* for key generation in 192 and 256 bits */
+#include <nettle/hmac.h> /* to check if custom hmac is supported */
 #include "sha-padlock.h"
 #endif
 #include "aes-padlock.h"
@@ -459,6 +460,7 @@ static void register_x86_padlock_crypto(unsigned capabilities)
 			gnutls_assert();
 		}
 
+#if defined(HAVE_LIBNETTLE) && defined(HMAC_SET_KEY)
 		ret = gnutls_crypto_single_mac_register(
 			GNUTLS_MAC_SHA1, 80, &_gnutls_hmac_sha_x86_ssse3, 0);
 		if (ret < 0)
@@ -473,6 +475,7 @@ static void register_x86_padlock_crypto(unsigned capabilities)
 			GNUTLS_MAC_SHA256, 80, &_gnutls_hmac_sha_x86_ssse3, 0);
 		if (ret < 0)
 			gnutls_assert();
+#endif
 
 		ret = gnutls_crypto_single_digest_register(
 			GNUTLS_DIG_SHA384, 80, &_gnutls_sha_x86_ssse3, 0);
@@ -483,6 +486,7 @@ static void register_x86_padlock_crypto(unsigned capabilities)
 			GNUTLS_DIG_SHA512, 80, &_gnutls_sha_x86_ssse3, 0);
 		if (ret < 0)
 			gnutls_assert();
+#if defined(HAVE_LIBNETTLE) && defined(HMAC_SET_KEY)
 		ret = gnutls_crypto_single_mac_register(
 			GNUTLS_MAC_SHA384, 80, &_gnutls_hmac_sha_x86_ssse3, 0);
 		if (ret < 0)
@@ -492,6 +496,7 @@ static void register_x86_padlock_crypto(unsigned capabilities)
 			GNUTLS_MAC_SHA512, 80, &_gnutls_hmac_sha_x86_ssse3, 0);
 		if (ret < 0)
 			gnutls_assert();
+#endif
 	}
 
 	if (check_optimized_aes()) {
@@ -692,6 +697,7 @@ static void register_x86_padlock_crypto(unsigned capabilities)
 				gnutls_assert();
 			}
 
+#if defined(HAVE_LIBNETTLE) && defined(HMAC_SET_KEY)
 			ret = gnutls_crypto_single_mac_register(
 				GNUTLS_MAC_SHA384, 80,
 				&_gnutls_hmac_sha_padlock, 0);
@@ -705,6 +711,7 @@ static void register_x86_padlock_crypto(unsigned capabilities)
 			if (ret < 0) {
 				gnutls_assert();
 			}
+#endif
 		}
 
 		ret = gnutls_crypto_single_digest_register(
@@ -725,6 +732,7 @@ static void register_x86_padlock_crypto(unsigned capabilities)
 			gnutls_assert();
 		}
 
+#if defined(HAVE_LIBNETTLE) && defined(HMAC_SET_KEY)
 		ret = gnutls_crypto_single_mac_register(
 			GNUTLS_MAC_SHA1, 90, &_gnutls_hmac_sha_padlock, 0);
 		if (ret < 0) {
@@ -738,6 +746,7 @@ static void register_x86_padlock_crypto(unsigned capabilities)
 		if (ret < 0) {
 			gnutls_assert();
 		}
+#endif
 	} else if (phe) {
 		/* Original padlock PHE. Does not support incremental operations.
 		 */
@@ -755,6 +764,7 @@ static void register_x86_padlock_crypto(unsigned capabilities)
 			gnutls_assert();
 		}
 
+#if defined(HAVE_LIBNETTLE) && defined(HMAC_SET_KEY)
 		ret = gnutls_crypto_single_mac_register(
 			GNUTLS_MAC_SHA1, 90, &_gnutls_hmac_sha_padlock_oneshot,
 			0);
@@ -768,6 +778,7 @@ static void register_x86_padlock_crypto(unsigned capabilities)
 		if (ret < 0) {
 			gnutls_assert();
 		}
+#endif
 	}
 #endif
 
@@ -910,6 +921,7 @@ static void register_x86_intel_crypto(unsigned capabilities)
 			gnutls_assert();
 		}
 
+#if defined(HAVE_LIBNETTLE) && defined(HMAC_SET_KEY)
 		ret = gnutls_crypto_single_mac_register(
 			GNUTLS_MAC_SHA1, 80, &_gnutls_hmac_sha_x86_ssse3, 0);
 		if (ret < 0)
@@ -924,6 +936,7 @@ static void register_x86_intel_crypto(unsigned capabilities)
 			GNUTLS_MAC_SHA256, 80, &_gnutls_hmac_sha_x86_ssse3, 0);
 		if (ret < 0)
 			gnutls_assert();
+#endif
 
 		ret = gnutls_crypto_single_digest_register(
 			GNUTLS_DIG_SHA384, 80, &_gnutls_sha_x86_ssse3, 0);
@@ -934,6 +947,7 @@ static void register_x86_intel_crypto(unsigned capabilities)
 			GNUTLS_DIG_SHA512, 80, &_gnutls_sha_x86_ssse3, 0);
 		if (ret < 0)
 			gnutls_assert();
+#if defined(HAVE_LIBNETTLE) && defined(HMAC_SET_KEY)
 		ret = gnutls_crypto_single_mac_register(
 			GNUTLS_MAC_SHA384, 80, &_gnutls_hmac_sha_x86_ssse3, 0);
 		if (ret < 0)
@@ -943,6 +957,7 @@ static void register_x86_intel_crypto(unsigned capabilities)
 			GNUTLS_MAC_SHA512, 80, &_gnutls_hmac_sha_x86_ssse3, 0);
 		if (ret < 0)
 			gnutls_assert();
+#endif
 	}
 
 	if (check_optimized_aes()) {
-- 
2.53.0

